tower-map/src/service.rs

```diff
@@ -0,0 +1,47 @@
+use futures::{IntoFuture, Poll};
+use futures::future::Future;
+
+use tower_service::Service;
+use super::Error;
+
+/// Map the reqeust from type A to type B
+#[derive(Debug)]
+pub struct MapRequest<S, M> {
+  /// the inner service
+  inner: S,
+  /// the function to convert from the upstream request type
+  /// to the inner request type
+  m: M,
+}
+
+impl<S, M> MapRequest<S, M> {
+  /// Create a new rate limiter
+  pub fn new(m: M, inner: S) -> Self {
+    MapRequest { m, inner }
+  }
+}
+
+impl<S, M, F, O, Request> Service<Request> for MapRequest<S, M>
+where
+  // Func(Request) -> Future<S::Request, ?>
+  M: FnMut(Request) -> F,
```

Comment:
with futures,  mapping is not asynchronous. AndThen is used for asynchronous mapping. should we follow this pattern? I.e. MapRequest would not be Async and we would have a different util for Async mapping.

Reply:
@carllerche i'll make it synchronous for now, that would make things a bit easier. We can as you say, have a different utility for async mapping.